### PR TITLE
lint `jwst/fits_generator` and `jwst/associations`

### PR DESCRIPTION
**Pull request by zacharyburnett**
`jwst/fits_generator` and `jwst/associations` are currently excluded from linting with `ruff` and therefore `pre-commit`

After merging this, when devs change files in those modules, `pre-commit` should lint their changes

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] ~added entry in `CHANGES.rst` within the relevant release section~
- [x] ~updated or added relevant tests~
- [x] ~updated relevant documentation~
- [ ] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [x] ~Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)~

### CODE REVIEW

**Review by melanieclarke — APPROVED** [review]

File: jwst/associations/lib/diff.py
Comment (zacharyburnett) [inline jwst/associations/lib/diff.py:393]: will this meaningfully affect the data munging?

File: jwst/associations/lib/rules_level2_base.py
Comment (zacharyburnett) [inline jwst/associations/lib/rules_level2_base.py:793]: I assume these `def`s are equivalent to the lambdas, though I'm not sure what the lambdas are doing here anyway. I can revert this if it causes issues

File: jwst/associations/lib/update_path.py
Comment (zacharyburnett) [inline jwst/associations/lib/update_path.py:48]: same as above

File: jwst/associations/lib/update_path.py
Comment (zacharyburnett) [inline jwst/associations/lib/update_path.py:121]: same as above

**Review by zacharyburnett — COMMENTED** [review]
lambdas changed to `def` to satisfy rule [`E731 `](https://docs.astral.sh/ruff/rules/lambda-assignment/); I can revert this change if it messes with things

File: jwst/fits_generator/objects.py
Comment (zacharyburnett) [inline jwst/fits_generator/objects.py:568]: should this variable have been used somewhere?

File: jwst/fits_generator/proposalparser.py
Comment (zacharyburnett) [inline jwst/fits_generator/proposalparser.py:52]: same as above

File: jwst/fits_generator/verifiers.py
Comment (zacharyburnett) [inline jwst/fits_generator/verifiers.py:98]: I assume this line is doing something with bit shifting so is needed

**Review by zacharyburnett — COMMENTED** [review]
unused variables

**Comment by zacharyburnett** [comment]
@melanieclarke I applied the Ruff check as best I could, but I have a few questions about lambdas and unused variables (in review comments)

**Comment by melanieclarke** [comment]
It looks like there are some unit test failures.  We'll also need to run regression tests for these changes, when the unit tests are passing.

I'll check out the lambdas and unused variables.

File: jwst/associations/lib/association_rules.py
Comment (melanieclarke) [inline jwst/associations/lib/association_rules.py:7]: I think this import is implicitly needed.

**Review by melanieclarke — COMMENTED** [review]
I think the lambdas are okay, but some of the `import *`s were functional and need fixing.  I pointed out a couple places where I can get more unit tests to pass with an extra import, but there is another test that's failing that looks import related (`jwst/associations/tests/test_registry.py::test_registry_match`).  I'm not sure which one is causing that failure.

File: jwst/fits_generator/objects.py
Comment (melanieclarke) [reply jwst/fits_generator/objects.py:568]: It looks to me like the function call is needed, but the return value is not.  It is probably equivalent to change the line to just `hdudef.describe(stream, state)`.

File: jwst/fits_generator/proposalparser.py
Comment (melanieclarke) [reply jwst/fits_generator/proposalparser.py:52]: From the git history, it doesn't look like that value has ever been used.  It can probably be removed.

File: jwst/fits_generator/verifiers.py
Comment (melanieclarke) [reply jwst/fits_generator/verifiers.py:98]: This whole function looks wrong - `value` is not defined anywhere, and neither is `error_collector`.  I hope and trust this isn't being used, so we should probably remove it.

File: jwst/fits_generator/objects.py
Comment (zacharyburnett) [reply jwst/fits_generator/objects.py:568]:
```suggestion
            hdudef.describe(stream, state) # noqa F841
```

File: jwst/fits_generator/proposalparser.py
Comment (zacharyburnett) [reply jwst/fits_generator/proposalparser.py:52]:
```suggestion
```

File: jwst/associations/lib/rules_level2b.py
Comment (zacharyburnett) [reply jwst/associations/lib/rules_level2b.py:18]:
```suggestion
from jwst.associations.lib.rules_level2_base import (
  AsnMixin_Lv2Image, AsnMixin_Lv2Spectral, AsnMixin_Lv2Nod, AsnMixin_Lv2Special, 
  DMSLevel2bBase, DMSAttrConstraint, Utility, 
  Constraint_Mode, Constraint_Base, Constraint_Background, Constraint_Single_Science, 
  Constraint_Image_Science, Constraint_Image_Nonscience, 
  Constraint_Special, Constraint_Spectral_Science, Constraint_Imprint, 
  Constraint_Target
)
```

File: jwst/associations/lib/rules_level3.py
Comment (zacharyburnett) [reply jwst/associations/lib/rules_level3.py:11]:
```suggestion
from jwst.associations.lib.rules_level3_base import (
  AsnMixin_Science, AsnMixin_AuxData, AsnMixin_Coronagraphy, AsnMixin_Spectrum, 
  DMS_Level3_Base, DMSAttrConstraint, 
  Constraint, SimpleConstraint, Constraint_Optical_Path, Constraint_Target, Constraint_Image, Constraint_MSA, Constraint_IFU
)
```

File: jwst/associations/__init__.py
Comment (zacharyburnett) [reply jwst/associations/__init__.py:27]: oh whoops, this should have been `E402`

File: jwst/fits_generator/verifiers.py
Comment (zacharyburnett) [reply jwst/fits_generator/verifiers.py:98]: is `fits_generator` used anywhere at all?

Comment (melanieclarke) [reply jwst/fits_generator/verifiers.py:98]: It looks like some parts of the module are called by scripts in jwst/scripts.  I can't say if those scripts are used, but clearly, these functions in template and verifiers are not!  
@tapastro - do you know more about the fits_generator module?

**Comment by melanieclarke** [comment]
> there is another test that's failing that looks import related (`jwst/associations/tests/test_registry.py::test_registry_match`). I'm not sure which one is causing that failure.

I think this one is because the imports in rules_level2b and rules_level3 are missing ASN_SCHEMA.

**Comment by zacharyburnett** [comment]
> > there is another test that's failing that looks import related (`jwst/associations/tests/test_registry.py::test_registry_match`). I'm not sure which one is causing that failure.
> 
> I think this one is because the imports in rules_level2b and rules_level3 are missing ASN_SCHEMA.

yep that fixed it!
```
❯ pytest -n auto jwst/associations/tests/test_registry.py
============================== test session starts ==============================
platform darwin -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0
crds_context: jwst_1253.pmap
rootdir: /Users/zburnett/projects/jwst
configfile: pyproject.toml
plugins: cov-5.0.0, asdf-3.3.0, ci_watson-0.7.0, doctestplus-1.2.1, jwst-1.15.2.dev14+gf6562101f, xdist-3.6.1, requests-mock-1.12.1
8 workers [6 items]     
......                                                                    [100%]
=============================== 6 passed in 3.44s ===============================
```

**Review by emolter — COMMENTED** [review]
Changes look good to me, but I don't see the regression test run linked anywhere even though the box is checked - did I miss it somewhere?  I approve this conditional on no regtest failures.

With regards to the use of `fits_generator`, I think it's a good idea to figure out whether and how it's used, but if we want to remove it (or parts of it) I assume there should probably be a separate JP ticket and we should ping some INS folks, i.e., it wouldn't have to hold this ticket up.

**Comment by zacharyburnett** [comment]
> Changes look good to me, but I don't see the regression test run linked anywhere even though the box is checked - did I miss it somewhere? I approve this conditional on no regtest failures.

whoops, I forgot about that, here's the regtest run now: https://github.com/spacetelescope/RegressionTests/actions/runs/9975534597

File: jwst/fits_generator/verifiers.py
Comment (melanieclarke) [reply jwst/fits_generator/verifiers.py:98]: Per discussion in this morning's tag up, the fits_generator module was mostly for generating test data, pre-flight, and is no longer used.  It should be deprecated and removed, but that's beyond the scope of this PR.  For now, we can just make the minimal changes to get it past the linter.

**Comment by emolter** [comment]
regtest failures look the same as the ones on main, i.e., related to the architecture differences when switching over to the new GitHub actions workflow.  So they are unrelated to this PR

**Review by emolter — APPROVED** [review]

**Review by melanieclarke — APPROVED** [review]

**Pull request merged**
